renku/service/errors.py

```diff
@@ -16,53 +16,675 @@
+
+    def __init__(self, exception=None, branch=ERROR_NOT_AVAILABLE, error_message=ERROR_NOT_AVAILABLE):
+        super().__init__(
+            userMessage=self.userMessage.format(branch=branch),
+            devMessage=self.devMessage.format(error_message=error_message),
+            exception=exception,
+        )
+
+
+class UserAnonymousError(ServiceError):
+    """The user must login in to use the target endpoint."""
+
+    code = SVC_ERROR_USER + 30
+    userMessage = "It's necessary to be authenticated to access the target data."
+    devMessage = "User identification is incorrect or missing."
+
+    def __init__(self, exception=None):
+        super().__init__(exception=exception)
+
+
+class UserMissingFieldError(ServiceError):
+    """The user has not provided an expected field."""
+
+    code = SVC_ERROR_USER + 40
+    userMessage = "The following field is required: {field}."
+    devMessage = "User did not provide the mandatory field '{field}'."
+
+    def __init__(self, exception=None, field=ERROR_NOT_AVAILABLE):
+        super().__init__(
+            userMessage=self.userMessage.format(field=field),
+            devMessage=self.devMessage.format(field=field),
+            exception=exception,
+        )
+
+
+class UserTemplateInvalidError(ServiceError):
+    """The provided URL doesn't lead to a valid template repository."""
+
+    code = SVC_ERROR_USER + 101
+    userMessage = "The target repository is not a valid RenkuLab template repository."
+    devMessage = "Target repository is not a valid template."
+
+    def __init__(self, exception=None):
+        super().__init__(exception=exception)
+
+
+class UserProjectCreationError(ServiceError):
+    """Error when creating a new project from a Renku template.
+
+    One (or more) project field is wrong.
+    """
+
+    code = SVC_ERROR_USER + 102
+    userMessage = "There is an error with a project field: {error_message}."
+    devMessage = "Project creation from a Renku template failed. The user provided wrong field(s): {error_message}."
+
+    def __init__(self, exception=None, error_message=ERROR_NOT_AVAILABLE):
+        super().__init__(
+            userMessage=self.userMessage.format(error_message=error_message),
+            devMessage=self.devMessage.format(error_message=error_message),
+            exception=exception,
+        )
+
+
+class UserNonRenkuProjectError(ServiceError):
+    """The target repository is valid but it is not a Renku project."""
+
+    code = SVC_ERROR_USER + 110
+    userMessage = "The target project is not a Renku project."
+    devMessage = "Cannot work on a non Renku project."
+
+    def __init__(self, exception=None):
+        super().__init__(exception=exception)
+
+
+class UserDatasetsMultipleImagesError(ServiceError):
+    """Multiple images have been set for the same dataset."""
+
+    code = SVC_ERROR_USER + 130
+    userMessage = "Cannot set multiple images for the dataset."
```

Comment:
This one is slightly off. You can set more than one images for a dataset (i.e. first one being the thumbnail and then some other images, maybe charts or something?). But each image needs a position(int), so they're ordered. And you can't have two images with the same position. So `[{url: img1.png, position: 1},{url: img2.png, position:1}]` gives this error, but `[{url: img1.png, position: 1},{url: img2.png, position:2}]` is perfectly fine.